Pulumi Challenge # 4 - Holiday Shopping (#2262)
pulumi/pulumi-hugo: themes/default/content/challenge/holiday-shopping.md

                )],
            ),
            "back": awsx.ecs.TaskDefinitionContainerDefinitionArgs(
                image=backend_image.image_uri,
                cpu=cpu,
                memory=memory,
                essential=True,
                port_mappings=[awsx.ecs.TaskDefinitionPortMappingArgs(
                    container_port=8000,
                    host_port=8000
                )],
                environment=[{
                    "name":"DATABASE_URL",
                    "value":Output.format("mongodb+srv://{0}:{1}@{2}", db_username, db_password,
                                Output.all(mongo_cluster.srv_address).apply(lambda v: v[0].split("//"))[1])
                }],
            ),
        }
    ),
    desired_count=1
)
```

#### Step 5. Validate endpoints

Run `pulumi up` and select yes to update the stack. Go to the `app url` export to validate that the endpoints of your deployed infrastructure is working. You can now add, delete, and check off items because the application is now able to connect to your newly created Atlas database.

![alt_text](/images/challenge/grocery_list.png "grocery list app")

### Congratulations

> Congratulations!. You've completed this Pulumi Challenge. If you want to receive special swag for completing the challenge (and to thank The Pulumipus), write a blog or post a quick video about it, and then tag us on social media or email us at [da@pulumi.com](mailto:da@pulumi.com).

Author: ```suggestion
Congratulations! You've completed this Pulumi Challenge. If you want to receive special swag for completing the challenge (and to thank The Pulumipus), write a blog or post a quick video about it, and then tag us on social media or email us at [da@pulumi.com](mailto:da@pulumi.com).
```

---
title: "Holiday Shopping"
layout: challenge/single
description: |
    Create a simple shopping list app using AWS Fargate and MongoDB Atlas
meta_desc: |
    Create a simple shopping list app using AWS Fargate and MongoDB Atlas
meta_image: /images/challenge/challenge-dec-meta.png
banner_image: /images/challenge/challenge-dec-banner.png
---

## Welcome to the Pulumi Challenge!

### Holiday Shopping List

The holidays are upon us and mere anarchy has been loosed upon our psyches. More specifically how do you keep track of everything you need to buy for gifts or food?!? Surely some salvation is at hand? Lucky for you, our friend, The Pulumipus, has created a simple to-do list application to bring order to your life. In this challenge, you will build this simple to-do list application that has a React frontend, Express backend, and a MongoDB database. This frontend and backend of the app run in Fargate with an Application Load Balancer (ALB) in front and the database is running in a MongoDB Atlas cluster.

![alt_text](/images/challenge/architecture_mern.png "mern arch")

> If you feel so inclined to thank The Pulumipus, please write a blog post or post a quick video about it. Tag us on social media or email us at [da@pulumi.com](mailto:da@pulumi.com), and we will send you some special swag.

Author: ```suggestion
After completing this challenge and if you feel so inclined to thank The Pulumipus, please write a blog post or post a quick video about it. Tag us on social media or email us at [da@pulumi.com](mailto:da@pulumi.com), and we will send you some special swag.
```

### Prerequisites

In order to complete this challenge, you'll need a couple things set up in advance. The entire challenge can be run on the free tiers of AWS and MongoDB Cloud.

- A [Pulumi account](https://app.pulumi.com/signup)
- The [Pulumi CLI](/docs/get-started/install/)
- [Python 3.9 or higher](https://www.python.org/downloads/)
- Docker
- An [AWS account](https://portal.aws.amazon.com/)
- The [AWS CLI](https://docs.aws.amazon.com/cli/latest/userguide/getting-started-install.html)
- A [MongoDB Cloud account](https://www.mongodb.com/cloud/atlas/signup)
- A [GitHub account](https://github.com/signup)

### Challenge

#### Step 1. Initialize a project

You will first clone the base application and then the rest of the steps will take you through building the Pulumi program.

```shell
git clone -b challenge https://github.com/aaronkao/atlas-fargate.git
```

Next, initialize the stack.

```shell
cd atlas-fargate
pulumi stack init dev
```

> You will also need to export your AWS and MongoDB Cloud credentials and your AWS region (e.g., us-west-2). To access your MongoDB Cloud credentials, go to the __Organization Access Manager__ panel and the __API Keys__ tab.

Reviewer: We need to tell them what to pick here, in case they're creating a new key (which they probably will be):

![image](https://user-images.githubusercontent.com/274700/205185275-7f8d83d5-38b3-4c40-b4d5-c6523d7e9338.png)

I believe you need at least Project Creator, as the others all appear to be read-only.

> You will also need to export your AWS and MongoDB Cloud credentials and your AWS region (e.g., us-west-2). To access your MongoDB Cloud credentials, go to the __Organization Access Manager__ panel and the __API Keys__ tab.

Author: ```suggestion
You will also need to export your AWS and MongoDB Cloud credentials and your AWS region (e.g., us-west-2). To access your MongoDB Cloud credentials, go to the __Organization Access Manager__ panel and the __API Keys__ tab. If you don't have any API keys, please click Create API Key and select _Organization Project Creator_ under the Organization Permissions.
```

memory = config.get_int("memory", 1024)

# Deploy an ECS Service on Fargate to host the application containers
service = awsx.ecs.FargateService(
    "grocery-service",
    cluster=cluster.arn,
    assign_public_ip=True,
    task_definition_args=awsx.ecs.FargateServiceTaskDefinitionArgs(
        containers={
            "front": awsx.ecs.TaskDefinitionContainerDefinitionArgs(
                image=frontend_image.image_uri,
                cpu=cpu,
                memory=memory,
                essential=True,
                port_mappings=[awsx.ecs.TaskDefinitionPortMappingArgs(
                    container_port=container_port,
                    target_group=lb.default_target_group,
                )],
            ),
            "back": awsx.ecs.TaskDefinitionContainerDefinitionArgs(
                image=backend_image.image_uri,
                cpu=cpu,
                memory=memory,
                essential=True,
                port_mappings=[awsx.ecs.TaskDefinitionPortMappingArgs(
                    container_port=8000,
                    host_port=8000
                )],
            ),
        }
    ),
    desired_count=1
)

# AWS export

> pulumi.export("app url", Output.concat("http://", lb.load_balancer.dns_name))

Reviewer: ```suggestion
pulumi.export("app_url", Output.concat("http://", lb.load_balancer.dns_name))
```

    comment="Open access for backend",
    project_id=mongo_project.id,)

# Create Free Tier cluster
mongo_cluster = mongodb.Cluster("mongo-cluster",
    backing_provider_name="AWS",
    project_id=mongo_project.id,
    provider_instance_size_name="M0",
    provider_name="TENANT",
    provider_region_name="US_WEST_2")

# Create Database user and give access to the cluster and database
mongo_user = mongodb.DatabaseUser("db_user",
    auth_database_name="admin",
    labels=[mongodb.DatabaseUserLabelArgs(
        key="project",
        value="pulumi",
    )],
    password=db_password,
    project_id=mongo_project.id,
    roles=[
        mongodb.DatabaseUserRoleArgs(
            database_name="grocery-list",
            role_name="readWrite",
        ),
        mongodb.DatabaseUserRoleArgs(
            database_name="admin",
            role_name="readAnyDatabase",
        ),
    ],
    scopes=[
        mongodb.DatabaseUserScopeArgs(
            # Extracts the cluster name to add to database scopes
            name=Output.all(mongo_cluster.srv_address).apply(lambda v: re.split("\.|\/\/", v[0])[1]),
            type="CLUSTER",
        ),
    ],
    username=db_username
)

> # MongoDB Atlas export
> pulumi.export("mongo connection string",
>     Output.format("mongodb+srv://{0}:{1}@{2}", db_username, db_password,
>         Output.all(mongo_cluster.srv_address).apply(lambda v: v[0].split("//"))[1])
> )
> ```

Reviewer: It doesn't look like this export is referenced -- should we be exporting it? The value is currently masked with `[secret]` anyway, so unless we explain that you need to `pulumi stack output --show-secrets` to see it, it won't be clear what it's there for. 

If we do want to export it though, it's generally better to use a contiguous string, just do you don't have to surround the values you want to read with quotes:

```suggestion
# MongoDB Atlas export
pulumi.export("mongo_connection_string",
    Output.format("mongodb+srv://{0}:{1}@{2}", db_username, db_password,
        Output.all(mongo_cluster.srv_address).apply(lambda v: v[0].split("//"))[1])
)
```
```

Author: This has just been removed since it isn't being used

Next, you need to set the config values for your MongoDB Cloud org, database username, and database password. The Organization ID is under the __Organization Settings__ panel in the MongoDB Cloud console. The database username and password values are self-selected and don't require you to look up anything in the MongoDB Cloud console.

```shell
pulumi config set orgID [value]
pulumi config set dbUser [value]
pulumi config set dbPassword [value] --secret
```

#### Step 4. Pass database URL as environment variable to backend container

Before you update the stack, you need to pass the URL of the database to the backend container so it can connect to it. Add the highlighted into the FargateService code.

> ```python {.line-numbers hl_lines=["32-37"]}

Reviewer: ```suggestion
```python {.line-numbers hl_lines=["32-36"]}
```

    "grocery-service",
    cluster=cluster.arn,
    assign_public_ip=True,
    task_definition_args=awsx.ecs.FargateServiceTaskDefinitionArgs(
        containers={
            "front": awsx.ecs.TaskDefinitionContainerDefinitionArgs(
                image=frontend_image.image_uri,
                cpu=cpu,
                memory=memory,
                essential=True,
                port_mappings=[awsx.ecs.TaskDefinitionPortMappingArgs(
                    container_port=container_port,
                    target_group=lb.default_target_group,
                )],
                environment=[{
                    # Unused unless running dev server
                    "name":"VITE_BACKEND_URL",
                    "value":"http://localhost:8000"
                },
                ],
            ),
            "back": awsx.ecs.TaskDefinitionContainerDefinitionArgs(
                image=backend_image.image_uri,
                cpu=cpu,
                memory=memory,
                essential=True,
                port_mappings=[awsx.ecs.TaskDefinitionPortMappingArgs(
                    container_port=8000,
                    host_port=8000
                )],

>                 environment=[{
>                     "name":"DATABASE_URL",
>                     "value":Output.format("mongodb+srv://{0}:{1}@{2}", db_username, db_password,
>                                 Output.all(mongo_cluster.srv_address).apply(lambda v: v[0].split("//"))[1])
>                 },
>                 ],

Reviewer: ```suggestion
                environment=[{
                    "name":"DATABASE_URL",
                    "value":Output.format("mongodb+srv://{0}:{1}@{2}", db_username, db_password,
                                Output.all(mongo_cluster.srv_address).apply(lambda v: v[0].split("//"))[1])
                }],
```

            ),
        }
    ),
    desired_count=1
)
```

#### Step 5. Validate endpoints

Run `pulumi up` and select yes to update the stack. Go to the `app url` export to validate that the endpoints of your deployed infrastructure is working. You can now add, delete, and check off items because the application is now able to connect to your newly created Atlas database.

> ![alt_text](/images/challenge/grocery_list.png "grocery list app")

Reviewer: ```suggestion
![A screenshot of the grocery list app](/images/challenge/grocery_list.png)
```

#### Congratulations

> Congratulations!. You've completed this Pulumi Challenge. If you'd like to tear down all of these resources, run `pulumi destroy -y`. Otherwise, have fun playing around with your infrastructure stack and add whatever you like! 🙂

Reviewer: ```suggestion
Congratulations! You've completed this Pulumi Challenge. If you'd like to tear down all of these resources, run `pulumi destroy -y`. Otherwise, have fun playing around with your infrastructure stack and add whatever you like! 🙂
```

---
title: "Holiday Shopping"
layout: challenge/single
description: |
    Create a simple shopping list app using AWS Fargate and MongoDB Atlas 
meta_desc: |
    Create a simple shopping list app using AWS Fargate and MongoDB Atlas
meta_image: /images/challenge/challenge_cta.png
---

> ## Welcome to the Pulumi Challenge!

Reviewer: Now that we have several challenges, should we link to other challenges from here somehow? Even if that's just to include a sort of breadcrumb that leads back to the challenge landing page?

Author: Do we need to do that in this markdown file or somewhere in the layouts?

Reviewer: In the layouts. I've added https://github.com/pulumi/devrel-team/issues/460 as a follow-up.

> <div class="flex flex-wrap md:mt-12">

Reviewer: We don't need all this HTML in here -- we can just use straight Markdown. (Earlier versions of this had imagery here, I think, which is probably where this comes from.)

  <div>
    <h3>Holiday Shopping List</h3>
    <p class="pr-12">

>       The holidays are upon us and mere anarchy has been loosed upon our psyche. More specifically how do you keep track of everything you need to buy for gifts or food?!? Surely some salvation is at hand? Lucky for you, our friend, The Pulumipus, has created a simple to-do list application to bring order to your life. In this challenge, you will build this simple to-do list application that has a React frontend, Express backend, and a MongoDB database. This frontend and backend of the app run in Fargate with an Application Load Balancer (ALB) in front and the database is running in a MongoDB Atlas cluster. When you're done, we'll send you a fancy piece of swag, just for this Challenge! If you feel so inclined to thank The Pulumipus, please write a blog post or post a quick video about it. Tag us on social media or email us at <a href=mailto:da@pulumi.com>da@pulumi.com</a>, and we will send you the special swag.

Reviewer: ```suggestion
      The holidays are upon us and mere anarchy has been loosed upon our psyches. More specifically how do you keep track of everything you need to buy for gifts or food?!? Surely some salvation is at hand? Lucky for you, our friend, The Pulumipus, has created a simple to-do list application to bring order to your life. In this challenge, you will build this simple to-do list application that has a React frontend, Express backend, and a MongoDB database. This frontend and backend of the app run in Fargate with an Application Load Balancer (ALB) in front and the database is running in a MongoDB Atlas cluster. When you're done, we'll send you a fancy piece of swag, just for this Challenge! If you feel so inclined to thank The Pulumipus, please write a blog post or post a quick video about it. Tag us on social media or email us at <a href=mailto:da@pulumi.com>da@pulumi.com</a>, and we will send you the special swag.
```

Reviewer: >When you're done, we'll send you a fancy piece of swag
>Tag us on social media or email us and we will send you the special swag

Are there two pieces of swag, then (one you generate yourself and an extra one for tagging or emailing? We should probably make this clear up front so there isn't any confusion about it.

    </p>
    <h3>Prerequisites</h3>

>     <p>In order to complete this challenge, you'll need a couple things set up in advance.</p>

Reviewer: Should mention Docker installed and running also, because we'll be building containers.

    <ul>
      <li>
        A <a href="https://app.pulumi.com/signup" target="_blank" rel="noopener noreferrer">Pulumi account</a>
      </li>
      <li>
        The <a href="/docs/get-started/install/" target="_blank" rel="noopener noreferrer">Pulumi CLI</a>
      </li>
      <li>
          <a href="https://www.python.org/downloads/">Python 3.9 or higher</a>
      </li>
      <li>
        An <a href="https://portal.aws.amazon.com/" target="_blank" rel="noopener noreferrer">AWS account</a>
      </li>
      <li>
        The <a href="https://docs.aws.amazon.com/cli/latest/userguide/getting-started-install.html" target="_blank" rel="noopener noreferrer">AWS CLI</a>
      </li>
      <li>

>         A <a href="https://www.mongodb.com/cloud/atlas/signup" target="_blank" rel="noopener noreferrer">MongoDB Cloud account</a>

Reviewer: We should probably make it super clear that a free account is fine here as well, since I wondered about that myself when we ran through this the other day.

      </li>
      <li>
        A <a href="https://github.com/signup" target="_blank" rel="noopener noreferrer">GitHub account</a>
      </li>
    </ul>
  </div>
</div>

### Challenge

#### Step 1. Initialize a project

You will first clone the base application and then the rest of the steps will take you through building the Pulumi program.

```shell
git clone -b challenge https://github.com/aaronkao/atlas-fargate.git
```

Next, initialize the stack.

```shell

> pulumi stack init dev

Reviewer: This doesn't quite work here, as the previous command leaves you at a level above where the Pulumi project file is. We should probably change to `atlas-fargate` after cloning.

# Deploy an ECS Service on Fargate to host the application containers
service = awsx.ecs.FargateService(
    "grocery-service",
    cluster=cluster.arn,
    task_definition_args=awsx.ecs.FargateServiceTaskDefinitionArgs(
        containers={
            "front": awsx.ecs.TaskDefinitionContainerDefinitionArgs(
                image=frontend_image.image_uri,
                cpu=cpu,
                memory=memory,
                essential=True,
                port_mappings=[awsx.ecs.TaskDefinitionPortMappingArgs(
                    container_port=container_port,
                    target_group=lb.default_target_group,
                )],
            ),
            "back": awsx.ecs.TaskDefinitionContainerDefinitionArgs(
                image=backend_image.image_uri,
                cpu=cpu,
                memory=memory,
                essential=True,
                port_mappings=[awsx.ecs.TaskDefinitionPortMappingArgs(
                    container_port=8000,
                    host_port=8000
                )],
            ),
        }  
    ),
    desired_count=1
)

# AWS export
pulumi.export("app url", Output.concat("http://", lb.load_balancer.dns_name))
```

> Run `pulumi up` and select yes to perform the update. Pulumi will output the `app url`. Go to the URL, and you will be presented with Pulumipus' Grocery List web application. You can try to add some items, but nothing will happen because there is no database connected yet. You will do that in the next step.

Reviewer: This first deploy took a _really_ long time for me (the Fargate service is still `creating` after 10 minutes), so we should probably mention this up front to set expectations.

---

Update -- the `up` failed on this:

```
  aws:ecs:Service (grocery-service):
    error: 1 error occurred:
        * creating urn:pulumi:dev::atlas-fargate::awsx:ecs:FargateService$aws:ecs/service:Service::grocery-service: 1 error occurred:
        * error waiting for ECS service (arn:aws:ecs:us-west-2:616138583583:service/cluster-1529ef7/grocery-service-bd35866) to reach steady state after creation: ResourceNotReady: exceeded wait attempts
```

Health-check related, presumably?

Author: So this just takes like 6 minutes for me, but I added a note that it might take some time. I also don't get a 503 error

Reviewer: Ok, I'll run through it again today then. (I actually don't think I've had a successful run of this at all on my own yet.)

#### Step 3. Create a MongoDB Atlas database

> The web app stores the grocery list items inside a MongoDB database. MongoDB Atlas is a fully managed cloud database that you will be using for this application. Add the below before the code from the previous steps in `__main__.py`. This code creates a MongoDB project, sets up the network access controls, creates a Free Tier cluster in `US_WEST_2` in AWS, and a database user that is scoped to the created cluster.

Reviewer: > Add the below before the code

We should probably say exactly where this code should go, like "above the ECR repo declaration" or something, just so that's clear.

```python
# MongoDB Atlas configs

> db_username = config.get("dbUser", "test-username")
> db_password = config.get_secret_object("dbPassword", "test-password")
> atlas_org_id = config.get("orgID")

Reviewer: ```suggestion
atlas_org_id = config.require("orgID")
db_username = config.require("dbUser")
db_password = config.require_secret("dbPassword")
```

Author: Should I actually propagate this code change back to the the main branch?

Reviewer: > Is there a way that I can override the default hero banner and replace it with this?
> 
> ![Pulumi Challenge Hero Banner_November_Option 3](https://user-images.githubusercontent.com/7718195/205140207-e9b903c5-bd55-4248-a072-339449b543a9.png)

Yes, but it'd take a layout change and a tweak to the frontmatter, assuming you want make overriding this image optional and fall back to what we have now. Want me to do that?

Reviewer: > Should I actually propagate this code change back to the the main branch?

I don't _think_ so, since this code only gets added during this walkthough ... right? Or is there a "finished" version of it somewhere?

Author: well i do like the fact there is a default value for dbuser and dbpassword - is that bad practice?

Reviewer: Ok, done -- building now.

Reviewer: >default value for dbuser and dbpassword - is that bad practice?

Sometimes I'll use a default username, but generally not a default password, because to the extent people use it, it's easier to guess. 😄 But I do think since we're seeing these values as part of the challenge code, it's fine to just expect users to enter them.

    auth_database_name="admin",
    labels=[mongodb.DatabaseUserLabelArgs(
        key="project",
        value="pulumi",
    )],
    password=db_password,
    project_id=mongo_project.id,
    roles=[
        mongodb.DatabaseUserRoleArgs(
            database_name="grocery-list",
            role_name="readWrite",
        ),
        mongodb.DatabaseUserRoleArgs(
            database_name="admin",
            role_name="readAnyDatabase",
        ),
    ],
    scopes=[
        mongodb.DatabaseUserScopeArgs(
            # Extracts the cluster name to add to database scopes
            name=Output.all(mongo_cluster.srv_address).apply(lambda v: re.split("\.|\/\/", v[0])[1]),
            type="CLUSTER",
        ),
    ],
    username=db_username
)

# MongoDB Atlas export
pulumi.export("mongo connection string",
    Output.format("mongodb+srv://{0}:{1}@{2}", db_username, db_password,
        Output.all(mongo_cluster.srv_address).apply(lambda v: v[0].split("//"))[1])
)
```

Next, you need to set the config values for your MongoDB Cloud org, database username, and database password.

```shell
pulumi config set orgID [value]
pulumi config set dbUser [value]

> pulumi config set dbPassword [value] --secret

Reviewer: If it's possible to link directly to the locations for these values in Atlas, we should do that just to keep folks from getting lost. Should also mention that the user and pass will be used to provision, so they don't need to be obtained from Atlas first.

I also remember getting the creds set up took a bit of figuring out -- we should add some instructions for this also (like what perms to use, where to get the keys from and how to set them, etc.).

Author: I can't link directly to it, but I did mention where to find it

                essential=True,
                port_mappings=[awsx.ecs.TaskDefinitionPortMappingArgs(
                    container_port=8000,
                    host_port=8000
                )],
                environment=[{
                    "name":"DATABASE_URL",
                    "value":Output.format("mongodb+srv://{0}:{1}@{2}", db_username, db_password,
                                Output.all(mongo_cluster.srv_address).apply(lambda v: v[0].split("//"))[1])
                },
                ],
            ),
        }
    ),
    desired_count=1
)
```

#### Step 5. Validate endpoints

Run `pulumi up` and select yes to update the stack. Go to the `app url` export to validate that the endpoints of your deployed infrastructure is working. You can now add, delete, and check off items because the application is now able to connect to your newly created Atlas database.

#### Step 6. Add Swag provider

something about swag

#### Congratulations

Congratulations!. You've completed this Pulumi Challenge. If you'd like to tear down all of these resources, run `pulumi destroy -y`. Otherwise, have fun playing around with your infrastructure stack and add whatever you like! :)

> If you want to receive additional swag for completing the challenge (and to thank The Pulumipus), write a blog or post a quick video about it, and then tag us on social media or email us at [da@pulumi.com](mailto:da@pulumi.com).

Reviewer: Okay cool, this answers my earlier question -- but it'd still be good to make this clearer up front I think.
